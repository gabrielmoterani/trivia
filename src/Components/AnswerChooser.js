import React from 'react';
import { PropTypes } from 'prop-types';
import { StyleSheet } from 'react-native';
import { Layout } from 'react-native-ui-kitten';
import AnswerButton from './AnswerButton';
import i18n from '../Locales/i18n';

const booleanAnswers = [i18n.t('game.true'), i18n.t('game.false')];

function AnswerChooser(props) {
  const { currentQuestion } = props;
  const isBoolean = currentQuestion.type === 'boolean';
  // TODO: shuffle answers if not boolean;
  const answers = isBoolean ? booleanAnswers : currentQuestion.answers;

  return (
    <Layout style={Style.answers}>
      {(answers || []).map((answer, i) =>
        <AnswerButton
          key={i}
          title={answer}
          value={answer}
          answer={props.currentAnswer.answer}
          correct={props.currentAnswer.correct}
          callback={props.onChoose}
        />
      )}
    </Layout>
  );
}

const Style = StyleSheet.create({
  answers: {
    flexDirection: 'row',
    marginTop: 50,
    paddingHorizontal: 25,
    alignSelf: 'center',
  },
});

AnswerChooser.propTypes = {
  currentQuestion: PropTypes.object.isRequired,
  currentAnswer: PropTypes.object.isRequired,
  onChoose: PropTypes.func.isRequired,
};

export default AnswerChooser;
