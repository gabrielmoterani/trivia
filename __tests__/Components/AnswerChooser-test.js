import React from 'react';
import { shallow } from 'enzyme';
import AnswerChooser from '../../src/Components/AnswerChooser';

const initialProps = {
  currentQuestion: {
    type: 'boolean',
    answers: ['True', 'False'],
  },
  currentAnswer: {},
  onChoose: jest.fn(),
};

const setup = (props = initialProps) => shallow(<AnswerChooser {...props} />);

describe('Test AnswerChooser component', () => {
  let wrapper = setup();

  it('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have at least two answer buttons', () => {
    const answerBtns = wrapper.find('AnswerButton');
    expect(answerBtns.length).toBeGreaterThanOrEqual(2);
  });

  it('should properly trigger action when click answer button', () => {
    const answerBtn = wrapper.find('AnswerButton').first();
    answerBtn.props().callback();

    expect(initialProps.onChoose).toHaveBeenCalled();
  });

  it('should properly render on "multiple type" question', () => {
    const currentQuestion = { type: 'multiple', answers: ['A', 'B', 'C', 'D'] };
    wrapper = setup({ ...initialProps, currentQuestion });
    expect(wrapper).toMatchSnapshot();
  });

  it('should have exact amount of answers options on "multiple type" question', () => {
    const currentQuestion = { type: 'multiple', answers: ['A', 'B', 'C', 'D'] };
    wrapper = setup({ ...initialProps, currentQuestion });
    const answerBtns = wrapper.find('AnswerButton');
    expect(answerBtns.length).toBe(4);
  });
});
