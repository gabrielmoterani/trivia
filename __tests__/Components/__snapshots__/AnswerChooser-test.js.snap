// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test AnswerChooser component should properly render on "multiple type" question 1`] = `
<LayoutComponent
  style={
    Object {
      "alignSelf": "center",
      "flexDirection": "row",
      "marginTop": 50,
      "paddingHorizontal": 25,
    }
  }
>
  <AnswerButton
    callback={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    key="0"
    title="A"
    value="A"
  />
  <AnswerButton
    callback={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    key="1"
    title="B"
    value="B"
  />
  <AnswerButton
    callback={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    key="2"
    title="C"
    value="C"
  />
  <AnswerButton
    callback={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    key="3"
    title="D"
    value="D"
  />
</LayoutComponent>
`;

exports[`Test AnswerChooser component should render properly 1`] = `
<LayoutComponent
  style={
    Object {
      "alignSelf": "center",
      "flexDirection": "row",
      "marginTop": 50,
      "paddingHorizontal": 25,
    }
  }
>
  <AnswerButton
    callback={[MockFunction]}
    key="0"
    title="True"
    value="True"
  />
  <AnswerButton
    callback={[MockFunction]}
    key="1"
    title="False"
    value="False"
  />
</LayoutComponent>
`;
